### PR TITLE
Give superpowers to arangsync to allow it to also replicate all SmartGraphCollections

### PR DESCRIPTION
**Pull request by mchacki**
### Scope & Purpose

In order to get DC-2-DC to work, we need to allow the ArangoSync process to set more (and dangerous) options when creating the Collections, it cannot use the GraphsAPI (like the user can), as it needs a bit more fine-grained control about the sharing being identical on source and target datacenter.

Tests for this are in ArangoSync repository.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [x] Backports
  - [x] Backport for 3.9: *TO BE DONE*
  - [x] Backport for 3.8: *For disjoint smart graphs*
  - [ ] Backport for 3.7: *save, no back port required.*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/VocBase/Methods/Collections.cpp [hunk 1]
```diff
@@ -1289,19 +1289,31 @@ arangodb::Result Collections::checksum(LogicalCollection& collection,
   return trx.finish(res);
 }
 
+/// @brief the list of collection attributes that are allowed by user-input
+/// this is to avoid retyping the same list twice
+#define COMMON_ALLOWED_COLLECTION_INPUT_ATTRIBUTES                            \
+  StaticStrings::DataSourceSystem, StaticStrings::DataSourceId, "keyOptions", \
```
Comment (hkernbach) [inline arangod/VocBase/Methods/Collections.cpp:1295]: "keyOptions" as StaticString?

File: arangod/VocBase/Methods/Collections.cpp [hunk 2]
```diff
@@ -1289,19 +1289,31 @@ arangodb::Result Collections::checksum(LogicalCollection& collection,
...
+      StaticStrings::WaitForSyncString, StaticStrings::CacheEnabled,          \
+      StaticStrings::ShardKeys, StaticStrings::NumberOfShards,                \
+      StaticStrings::DistributeShardsLike, "avoidServers",                    \
```
Comment (hkernbach) [inline arangod/VocBase/Methods/Collections.cpp:1298]: "avoidServers" as StaticString?

File: arangod/VocBase/Methods/Collections.cpp [hunk 3]
```diff
@@ -1289,19 +1289,31 @@ arangodb::Result Collections::checksum(LogicalCollection& collection,
...
+      StaticStrings::IsSmart, StaticStrings::ShardingStrategy,                \
+      StaticStrings::GraphSmartGraphAttribute, StaticStrings::Schema,         \
+      StaticStrings::SmartJoinAttribute, StaticStrings::ReplicationFactor,    \
+      StaticStrings::MinReplicationFactor, /* deprecated */                   \
+      StaticStrings::WriteConcern, "servers"
```
Comment (hkernbach) [inline arangod/VocBase/Methods/Collections.cpp:1303]: "servers" as StaticString?

File: arangod/VocBase/Methods/Collections.cpp [hunk 4]
```diff
@@ -1289,19 +1289,31 @@ arangodb::Result Collections::checksum(LogicalCollection& collection,
...
+/// this is to avoid retyping the same list twice
+#define COMMON_ALLOWED_COLLECTION_INPUT_ATTRIBUTES                            \
+  StaticStrings::DataSourceSystem, StaticStrings::DataSourceId, "keyOptions", \
+      StaticStrings::WaitForSyncString, StaticStrings::CacheEnabled,          \
+      StaticStrings::ShardKeys, StaticStrings::NumberOfShards,                \
+      StaticStrings::DistributeShardsLike, "avoidServers",                    \
+      StaticStrings::IsSmart, StaticStrings::ShardingStrategy,                \
+      StaticStrings::GraphSmartGraphAttribute, StaticStrings::Schema,         \
+      StaticStrings::SmartJoinAttribute, StaticStrings::ReplicationFactor,    \
+      StaticStrings::MinReplicationFactor, /* deprecated */                   \
```
Comment (hkernbach) [inline arangod/VocBase/Methods/Collections.cpp:1302]: Can we remove "MinReplicationFactor"?

File: arangod/VocBase/Methods/Collections.cpp [hunk 5]
```diff
@@ -1289,19 +1289,31 @@ arangodb::Result Collections::checksum(LogicalCollection& collection,
...
+      StaticStrings::WriteConcern, "servers"
+
 arangodb::velocypack::Builder Collections::filterInput(
-    arangodb::velocypack::Slice properties) {
+    arangodb::velocypack::Slice properties, bool allowDC2DCAttributes) {
+#ifdef USE_ENTERPRISE
+  if (allowDC2DCAttributes) {
```
Comment (hkernbach) [inline arangod/VocBase/Methods/Collections.cpp:1308]: Probably, we could add a TRI_ASSERT for cluster check here?

**Review by hkernbach — APPROVED** [review]
Technically LGTM, some (optional) code suggestions added.

File: arangod/VocBase/Methods/Collections.cpp (hunk 3 above)
Comment (mchacki) [reply arangod/VocBase/Methods/Collections.cpp:1303]: Static Strings are good point, but as this shall be a bug-fix I would do this in a different refactoring PR.

File: arangod/VocBase/Methods/Collections.cpp (hunk 5 above)
Comment (mchacki) [reply arangod/VocBase/Methods/Collections.cpp:1308]: I think for now yes, DC-2-DC only replicates clusters.
However I think there is no technical reason for this piece to only work in Cluster environment, so I would not like to include the ASSERT here, it is required in other places on DC2DC, where it matters that we are on a cluster.

**Pull request merged**